logstash-core/src/main/java/org/logstash/Javafier.java

```diff
@@ -8,63 +10,67 @@
      */
-    private Javafier(){}
+    private Javafier() {
+    }
 
     public static Object deep(Object o) {
-        if (o instanceof RubyString) {
-            return o.toString();
-        }
-        if (o instanceof String || o instanceof Float || o instanceof Double ||
-            o instanceof Long || o instanceof Integer || o instanceof Boolean ||
-            o instanceof Timestamp) {
-            return o;
-        }
-        if (o instanceof RubyFloat) {
-            return ((RubyFloat) o).getDoubleValue();
-        }
-        if (o instanceof JrubyTimestampExtLibrary.RubyTimestamp) {
-            return ((JrubyTimestampExtLibrary.RubyTimestamp) o).getTimestamp();
-        }
-        if (o instanceof RubyBoolean) {
-            return ((RubyBoolean) o).isTrue();
-        }
-        if (o instanceof RubyFixnum) {
-            return ((RubyFixnum) o).getLongValue();
-        }
-        if (o instanceof BiValue) {
-            return ((BiValue)o).javaValue();
-        } else if(o instanceof ConvertedMap) {
-            return ((ConvertedMap) o).unconvert();
-        }  else if(o instanceof ConvertedList) {
-            return ((ConvertedList) o).unconvert();
-        } else {
-            return fallback(o);
+        final Class<?> cls = o.getClass();
+        final Valuefier.Converter converter = CONVERTER_MAP.get(cls);
+        if (converter != null) {
+            return converter.convert(o);
         }
+        return fallbackConvert(o, cls);
     }
 
-    /**
-     * Cold path of {@link Javafier#deep(Object)}.
-     * We assume that we never get an input that is neither {@link ConvertedMap}, {@link ConvertedList}
-     * nor {@link BiValue}, but fallback to attempting to create a {@link BiValue} from the input
-     * before converting to a Java type.
-     * @param o Know to not be an expected type in {@link Javafier#deep(Object)}.
-     * @return Input converted to Java type
-     */
-    private static Object fallback(final Object o) {
-        try {
-            return BiValues.newBiValue(o).javaValue();
-        } catch (IllegalArgumentException e) {
-            final Class<?> cls = o.getClass();
-            throw new IllegalArgumentException(String.format(ERR_TEMPLATE, cls.getName(), cls.getSimpleName()));
+    private static Object fallbackConvert(final Object o, final Class<?> cls) {
+        for (final Map.Entry<Class<?>, Valuefier.Converter> entry : CONVERTER_MAP.entrySet()) {
+            if (entry.getKey().isAssignableFrom(cls)) {
+                final Valuefier.Converter found = entry.getValue();
+                CONVERTER_MAP.put(cls, found);
+                return found.convert(o);
+            }
         }
+        CONVERTER_MAP.put(cls, BIVALUES_CONVERTER);
+        return BIVALUES_CONVERTER.convert(o);
+    }
+
+    private static Map<Class<?>, Valuefier.Converter> initConverters() {
+        final Map<Class<?>, Valuefier.Converter> converters =
+            new ConcurrentHashMap<>(50, 0.2F, 1);
```

Comment:
More curious then anything...but since this is initialized in static scope, and only read, could this be a non regular `HashMap` and/or `Unmodifiable` ?

Reply:
@jakelandis unfortunately not because of the fallback that might modify this https://github.com/elastic/logstash/pull/8158/files#diff-46a0157ada034d162b510a0c9f9f03f9R38

... that said we could try an `IdentityHash` with copy-on-write at some point, but I don't think it makes much of a difference here.